src/filters/renderer/VideoRenderers/PixelShaderCache.cpp

```diff
@@ -0,0 +1,270 @@
+        HRESULT res = m_pD3DDev->CreatePixelShader(pixelShader->value.m_p, ppPixelShader);
+
+        if (FAILED(res)) {
+            DeleteCache(hash);
+        } else {
+            CString file_path;
+            if (GetCacheFilePath(file_path, hash)) {
+                TouchFile(file_path);
+            }
+        }
+        return res;
+    }
+
+    return E_FAIL;
+}
+
+void CPixelShaderCache::SavePixelShader(
+    LPCSTR pProfile,
+    LPCSTR pSourceData,
+    SIZE_T SourceDataSize,
+    LPCSTR pSourceName,
+    void* pCompiledData,
+    SIZE_T CompiledDataSize)
+{
+    if (!IsEnabled() || (CompiledDataSize > m_CachedShaderMaxSize)) {
+        return;
+    }
+
+    CString cacheFilePath;
+    if (GetCacheFilePath(cacheFilePath, pProfile, pSourceData, SourceDataSize, pSourceName)) {
+        if (FILE* fp = _tfsopen(cacheFilePath, _T("wt"), _SH_SECURE)) {
+            bool res = fwrite(&m_Version, sizeof(m_Version), 1, fp) == 1
+                       && fwrite(&CompiledDataSize, sizeof(CompiledDataSize), 1, fp) == 1
+                       && fwrite(pCompiledData, CompiledDataSize, 1, fp) == 1;
+
+            fclose(fp);
+
+            if (!res) {
+                DeleteFile(cacheFilePath);
+            }
+        }
+    }
+}
+
+void CPixelShaderCache::LoadCache()
+{
+    CString cacheFolder;
+    if (GetCacheFolder(cacheFolder)) {
+        CFileFind finder;
+        BOOL working = finder.FindFile(cacheFolder + _T("\\*"));
+
+        while (working) {
+            working = finder.FindNextFile();
+            if (!finder.IsDirectory() && !finder.IsDots()) {
+                if (!LoadCache(finder.GetFileName(), finder.GetFilePath())) {
+                    DeleteFile(finder.GetFilePath());
+                }
+            }
+        }
+    }
+}
+
+void CPixelShaderCache::DestroyCache()
+{
+    CString cacheFolder;
+    if (GetCacheFolder(cacheFolder)) {
+        CFileFind finder;
+        BOOL working = finder.FindFile(cacheFolder + _T("\\*"));
+
+        while (working) {
+            working = finder.FindNextFile();
+            if (!finder.IsDirectory() && !finder.IsDots()) {
+                DeleteFile(finder.GetFilePath());
+            }
+        }
+    }
+}
+
+bool CPixelShaderCache::IsEnabled() const
+{
+    return GetRenderersSettings().m_AdvRendSets.bCacheShaders;
+}
+
+void CPixelShaderCache::TouchFile(const CString& FileName) const
+{
+    HANDLE file_handle = CreateFile(FileName, GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                    NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, 0);
+
+    if (file_handle != INVALID_HANDLE_VALUE) {
+        SYSTEMTIME current_time;
+        GetSystemTime(&current_time);
+
+        FILETIME file_time;
+        if (SystemTimeToFileTime(&current_time, &file_time)) {
+            SetFileTime(file_handle, NULL, NULL, &file_time);
+        }
+
+        CloseHandle(file_handle);
+    }
+}
+
+bool CPixelShaderCache::IsFileOutdated(const CString& FileName) const
```

Comment:
This whole method could be greatly simplified by using `CFile` class with `GetStatus` and comparing `CTime`.